### ISSUE TITLE
Etherpad errors from Firefox iframe

### ISSUE THREAD

**Issue by rcmurray** [ether/etherpad-lite#5021]
**Describe the bug**
I'm seeing errors *** within Etherpad *** when I access Etherpad from an iframe in Firefox 88.0 (64-bit) on Mac. (These are Etherpad errors rather than access issues from Firefox.) The same issues do not occur from Chrome 90.0.4430.93 (Official Build) (x86_64) on Mac or Safari 14.0.3 (14610.4.3.1.7) on Mac. Moreover, the same URLs when launched independently (not within an iframe) from Firefox appear fine, although I do see an error in the Etherpad log. 

**To Reproduce**
1. Go to 'https://bazaar.lti.cs.cmu.edu/bazaar/login?roomName=climatechange&roomId=52715&id=1&username=Robbie&html=climate_change'
2. Click on the 'Joint Response' tab label toward the upper left. 
3. See error display within the iframe. 
4. Go to 'https://collab.lti.cs.cmu.edu:9001/p/jclimatechange52715?showChat=false&noColors=false', which is the same URL as the iframe. 
5. Page displays correctly.
6. Return to page at 'https://bazaar.lti.cs.cmu.edu/bazaar/login?roomName=climatechange&roomId=52715&id=1&username=Robbie&html=climate_change'.
7. Click on the 'Joint Response' tab label toward the upper left. 
8. See a second error displayed within the iframe. (I suspect this error is related to the page display in step 5.) 

**Expected behavior**
The content of the page shown in step (5) above should be shown in the iframe at step (3) above. 

**Screenshots**

1. 'Firefox from iframe - error 1.png': The error shown in step (3) of the procedure to reproduce.
<img width="1331" alt="Firefox from iframe - error 1" src="https://user-images.githubusercontent.com/55146439/116936751-43c30d80-ac36-11eb-853c-e76518d2967b.png">
 
2. 'Firefox direct - no apparent error.png': The display in step (5) of the procedure to reproduce. 
<img width="1331" alt="Firefox direct - no apparent error" src="https://user-images.githubusercontent.com/55146439/116936782-4b82b200-ac36-11eb-9fc9-4f50a43cd816.png">

3. 'Firefox from iframe - error 2.png': The error shown in step (8) of the procedure to reproduce. 
![Firefox iframe error 2](https://user-images.githubusercontent.com/55146439/116936947-808f0480-ac36-11eb-9cf0-0b7087ec246b.png)

**Server (please complete the following information):**
 - Etherpad version: 1.8.13
 - OS: [e.g., Ubuntu 20.04]: Ubuntu 18.04
 - Node.js version (`node --version`): v14.16.0
 - npm version (`npm --version`): 6.14.11

**Desktop (please complete the following information):**
 - OS: [e.g. iOS]: iOS v10.14.6
 - Browser [e.g. chrome, safari]: Firefox
 - Version [e.g. 22]: 88.0 (64-bit)

**Smartphone (please complete the following information):**. 
 - Device: [e.g. iPhone6]:  N/A
 - OS: [e.g. iOS8.1]:  N/A
 - Browser [e.g. stock browser, safari]:  N/A
 - Version [e.g. 22]:  N/A

**Additional context**

I'm attaching two versions of the same log, first in .txt format and then converted to .md. I manually added spaces to the log, lines 313-321 of the .txt version, between (a) the events in reproduction steps 1-3, and (b) the events in reproduction steps 4-8.

1. Etherpad log - from Firefox iframe and Firefox direct.txt
[Etherpad log - from Firefox iframe and Firefox direct.txt](https://github.com/ether/etherpad-lite/files/6418016/Etherpad.log.-.from.Firefox.iframe.and.Firefox.direct.txt)

2. firefox-log.md
[firefox-log.md](https://github.com/ether/etherpad-lite/files/6418021/firefox-log.md)

**Comment by rhansen** [comment]
I am unable to reproduce. I don't see a "Joint Response" tab when I visit https://bazaar.lti.cs.cmu.edu:9001/p/jclimatechange52715?showChat=false&noColors=false (from your step 1) and it looks like it isn't in an iframe.

**Comment by rcmurray** [comment]
I am so sorry. I pasted in the wrong URL for steps (1) and (6). I corrected them above to 'https://bazaar.lti.cs.cmu.edu/bazaar/login?roomName=climatechange&roomId=52715&id=1&username=Robbie&html=climate_change'.

Thank you for such a speedy response!

**Comment by webzwo0i** [comment]
Can you test if the following works:
load your page
click "joint response"
right click next to the editbar (the icons at the top of etherpad)
choose "this frame"
click "reload frame"

The second error you posted is kind of misleading. Etherpad starts an reconnection loop because of the first error. I took a look and it seems the error is in https://github.com/ether/etherpad-lite/blob/develop/src/static/js/caretPosition.js#L197
This seems to return `null` - I don't understand it though. Might be related to the pad not being displayed during loading. Just to be sure, can you add the below lines after the `const selection = window.getSelection();` line please?
```
console.log(`selection is ${selection} and window.getSelection is ${window.getSelection} and window is ${window}`)
```

**Comment by rcmurray** [comment]
Will do. Thanks for looking at this.

**Comment by rcmurray** [comment]
Doing "reload frame" worked -- see screenshot below. 

I added the logging line to caretPosition.js but I don't see it in the console log with logging set to 'Debug'. Do I need to do a compilation step first. Attached are 
1. The screenshot mentioned above.
2. The log. This time it includes only the iframe access with "reload frame" (not direct access outside of an iframe).
3. The updated file 'caretPosition.js' with the ending '.txt' added so that this comment utility will et me attach it.

<img width="1331" alt="Firefox from iframe - after reload frame" src="https://user-images.githubusercontent.com/55146439/117070054-40de2080-acfb-11eb-90e1-c4cdc24be2d1.png">

[Etherpad log - from Firefox iframe with reload frame and extra line in log.txt](https://github.com/ether/etherpad-lite/files/6423948/Etherpad.log.-.from.Firefox.iframe.with.reload.frame.and.extra.line.in.log.txt)


[caretPosition.js.txt](https://github.com/ether/etherpad-lite/files/6423967/caretPosition.js.txt)

**Comment by rcmurray** [comment]
Oops. I didn't see the line "and change the if (selection.rangeCount... line to
if (selection && selection.rangeCount > 0) {". I'll do that now and post again.

**Comment by rcmurray** [comment]
With the additional line added to caretPosition.js (again, sorry), Etherpad loaded correctly within the Firefox iframe with no "reload page" required! Attaching
1. Screenshot
2. Log

Is this the fix? What is the process going forward? (E.g., should I just change this line locally, and if so, how will I know when there is an official fix?)

Thanks very much for your help.

<img width="1331" alt="Firefox from iframe - correct" src="https://user-images.githubusercontent.com/55146439/117071574-45a3d400-acfd-11eb-9169-45cd879f63e0.png">

[Etherpad log - from Firefox iframe with caretPosition additions.txt](https://github.com/ether/etherpad-lite/files/6424034/Etherpad.log.-.from.Firefox.iframe.with.caretPosition.additions.txt)

**Comment by webzwo0i** [comment]
No need to apologize :-) Thanks for reporting this issue!
https://github.com/ether/etherpad-lite/pull/5023 is the proposed fix. If you want, reset your changes and checkout the PR.

**Comment by rhansen** [comment]
Merged #5023
